Arm/Insts/DPI/Add_sub_imm.lean

```diff
@@ -33,7 +33,9 @@ def exec_add_sub_imm (inst : Add_sub_imm_cls) (s : ArmState) : ArmState :=
     -- State Updates
     let s'            := write_pc ((read_pc s) + 4#64) s
     let s'            := if setflags then write_pstate pstate s' else s'
-    let s'            := write_gpr datasize inst.Rd result s'
+    -- When Rd == 31, the instruction is the alias CMP <Xn|XSP>,#<imm>[,<shift>]
+    -- This instruction writes out result to XZR
+    let s'            := write_gpr_zr datasize inst.Rd result s'
```

Comment:
The comment at L36 is a little misleading because the alias situation is different for each of the instructions being modeled here, and not all of them are only when Rd == 31:
1) ADD (imm) is used by the alias MOV (to/from SP): https://developer.arm.com/documentation/ddi0602/2022-03/Base-Instructions/ADD--immediate---Add--immediate--?lang=en
2) ADDS (imm) is used by the alias CMN (imm): https://developer.arm.com/documentation/ddi0602/2022-03/Base-Instructions/ADDS--immediate---Add--immediate---setting-flags-?lang=en
3) SUB (imm) is used by no alias (https://developer.arm.com/documentation/ddi0602/2022-03/Base-Instructions/SUB--immediate---Subtract--immediate--?lang=en)
4) SUBS (imm) is used by the CMP (imm) alias (https://developer.arm.com/documentation/ddi0602/2022-03/Base-Instructions/SUBS--immediate---Subtract--immediate---setting-flags-?lang=en)

Also, these instructions write to the Stack Pointer, which is the default behavior of `write_gpr` when called with register index 31 -- `write_gpr_zr` essentially throws out the result when register index is 31. I think `write_gpr` should be used here instead of `write_gpr_zr`.

Reply:
This change is added for the last case in 4. It says in the document that 

"CMP <Xn|SP>, #< imm >{, < shift >}

is equivalent to

SUBS XZR, <Xn|SP>, #< imm > {, < shift >}"
which writes out to the XZR register.

I didn't realize there are other alias that might collide with this one though.

Reply:
It seems like ADDS (like SUBS) also uses XZR: https://developer.arm.com/documentation/ddi0602/2022-03/Base-Instructions/ADDS--immediate---Add--immediate---setting-flags-?lang=en

Reply:
Oh, actually, I think we can just follow the specification:
```
if d == 31 && !setflags then
    SP[] = result;
else
    X[d] = result;
```
I believe in the else case, it would write to XZR when d == 31.